File: app/src/main/java/com/qimian233/ztool/hook/modules/mobiledesktop/BypassShareWarningHook.java
package com.qimian233.ztool.hook.modules.mobiledesktop;

import android.content.Context;

import com.qimian233.ztool.hook.base.BaseHookModule;
import com.qimian233.ztool.hook.base.DexKitHelper;

import io.github.libxposed.api.XposedModuleInterface;

import org.luckypray.dexkit.DexKitBridge;
import org.luckypray.dexkit.query.FindClass;
import org.luckypray.dexkit.query.FindMethod;
import org.luckypray.dexkit.query.matchers.ClassMatcher;
import org.luckypray.dexkit.query.matchers.MethodMatcher;
import org.luckypray.dexkit.query.matchers.MethodsMatcher;
import org.luckypray.dexkit.result.ClassData;
import org.luckypray.dexkit.result.MethodData;

import java.lang.reflect.Method;
import java.util.List;

/**
 * 绕过超级互联分享警告弹窗的 Hook。
 * <p>
 * 使用 DEXKit 通过方法签名（参数类型+返回类型）动态匹配混淆后的方法名，
 * 不再依赖硬编码的单字母名称。
 * </p>
 */
public class BypassShareWarningHook extends BaseHookModule {

    private static final String TARGET_PACKAGE = "com.motorola.mobiledesktop";
    private static final String TARGET_CLASS = "com.motorola.readyfor.tile.BaseFileUnionTile";
    private static final String DIALOG_CLASS =
            "com.motorola.readyfor.common.dialog.ActionNoticeCommonDialogActivity";
    private static final String MANAGER_PKG = "com.motorola.mobiledesktop.manager";
    private static final String PREFS_NAME = "moto_ble_preference";
    private static final String PREF_KEY = "file_union_transfer_switch";

    public BypassShareWarningHook() {}

    @Override
    public String getModuleName() {
        return "bypass_share_warning";
    }

    @Override
    public String[] getTargetPackages() {
        return new String[]{TARGET_PACKAGE};
    }

    @Override
    public void handleLoadPackage(XposedModuleInterface.PackageLoadedParam param) throws Throwable {
        ClassLoader classLoader = param.getDefaultClassLoader();

        // ── DEXKit：预解析管理器类和方法 ─────────────────────────────
        DexKitBridge bridge = DexKitHelper.INSTANCE.getBridgeForClass(
                classLoader, TARGET_CLASS);

        String managerClassName = null;
        String managerFactoryMethodName = null;   // static factory: (Context) → manager
        String managerSetMethodName = null;        // instance: (boolean) → void

        if (bridge != null) {
            try {
                // 1. 查找管理器类：在 manager 包中找一个有 static (Context) 方法和 (boolean)void 方法的类
                ClassData managerData = bridge.findClass(FindClass.create()
                        .searchPackages(MANAGER_PKG)
                        .matcher(ClassMatcher.create()
                                .methods(MethodsMatcher.create()
                                        .add(MethodMatcher.create()
                                                .modifiers(java.lang.reflect.Modifier.STATIC
                                                        | java.lang.reflect.Modifier.PUBLIC)
                                                .paramTypes("android.content.Context"))
                                        .add(MethodMatcher.create()
                                                .paramTypes("boolean")
                                                .returnType("void"))
                                )
                        )
                ).singleOrNull();

                if (managerData != null) {
                    managerClassName = managerData.getName();
                    List<MethodData> methods = managerData.getMethods();
                    for (MethodData md : methods) {
                        List<String> params = md.getParamTypeNames();
                        if (params.size() == 1 && "boolean".equals(params.get(0)) && "void".equals(md.getReturnTypeName())) {
                            managerSetMethodName = md.getName();
                        } else if (params.size() == 1 && "android.content.Context".equals(params.get(0)) && !"void".equals(md.getReturnTypeName())) {
                            managerFactoryMethodName = md.getName();
                        }
                    }
                }

                // 回退硬编码名称
                if (managerClassName == null) managerClassName = MANAGER_PKG + ".c0";
                if (managerFactoryMethodName == null) managerFactoryMethodName = "l";
                if (managerSetMethodName == null) managerSetMethodName = "z";

            } catch (Throwable dexKitError) {
                logError("DEXKit method discovery failed, using hardcoded names", dexKitError);
                managerClassName = MANAGER_PKG + ".c0";
                managerFactoryMethodName = "l";
                managerSetMethodName = "z";
            }
        } else {
            managerClassName = MANAGER_PKG + ".c0";
            managerFactoryMethodName = "l";
            managerSetMethodName = "z";
        }

        final String finalManagerClass = managerClassName;
        final String finalFactoryMethod = managerFactoryMethodName;
        final String finalSetMethod = managerSetMethodName;

        try {
            // ── Hook 1: 磁贴点击 ───────────────────────────────────
            Class<?> baseFileUnionTileClass = classLoader.loadClass(TARGET_CLASS);
            Method onClickMethod = baseFileUnionTileClass.getDeclaredMethod("onClick");
            this.xposed.hook(onClickMethod).intercept(chain -> {
                Object tile = chain.getThisObject();
                Context context = getContext(tile);
                if (context == null) {
                    return chain.proceed();
                }

                boolean enabled = isNearbyShareEnabled(context);
                log("IsNearbyShareEnabled: " + enabled);
                if (enabled) {
                    log("Nearby share already enabled, keep original disable flow.");
                    return chain.proceed();
                }

                setNearbyShareEnabled(tile, context, classLoader,
                        finalManagerClass, finalFactoryMethod, finalSetMethod);
                log("Bypassed warning and enabled nearby share directly.");
                return null;
            });
            log("Installed hook for BaseFileUnionTile.onClick");
        } catch (Throwable t) {
            logError("Failed to hook BaseFileUnionTile.onClick", t);
        }

        try {
            // ── Hook 2: 通用弹窗场景 ─────────────────────────────────
            Class<?> actionNoticeClass = classLoader.loadClass(DIALOG_CLASS);

            // 通过 DEXKit 动态查找 p() 方法 — 无参 void 方法
            String pMethodName = "p"; // 默认回退
            if (bridge != null) {
                try {
                    MethodData md = bridge.findMethod(FindMethod.create()
                            .searchPackages(TARGET_PACKAGE)
                            .matcher(MethodMatcher.create()
                                    .paramTypes()
                                    .returnType("void")
                                    .declaredClass(DIALOG_CLASS)
                            )
                    ).singleOrNull();
                    if (md != null) pMethodName = md.getName();
                } catch (Throwable ignored) {}
            }
            final String finalPMethodName = pMethodName;

            Method pMethod = actionNoticeClass.getDeclaredMethod(finalPMethodName);
            this.xposed.hook(pMethod).intercept(chain -> {
                Object myObject = chain.getThisObject();
                Context context = getContext(myObject);

                Class<?> managerClass = classLoader.loadClass(finalManagerClass);
                Method lMethod = managerClass.getDeclaredMethod(finalFactoryMethod, Context.class);
                Object manager = lMethod.invoke(null, context);
                if (manager == null) {
                    log("Unable to get manager!");
                    return null;
                }
                Method zMethod = manager.getClass().getDeclaredMethod(finalSetMethod, boolean.class);
                zMethod.setAccessible(true);
                zMethod.invoke(manager, true);
                return null;
            });
            log("Installed hook for dialog activity method: " + finalPMethodName);
        } catch (Exception e) {
            logError("Failed to hook createAndStartExposureWarnIntent: ", e);
        }
    }

    private Context getContext(Object tile) {
        try {
            Method getApplicationContextMethod = tile.getClass().getMethod("getApplicationContext");
            Object context = getApplicationContextMethod.invoke(tile);
            if (context instanceof Context) {
                return (Context) context;
            }
        } catch (Throwable ignored) {
        }
        return null;
    }

    private boolean isNearbyShareEnabled(Context context) {
        try {
            return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                    .getBoolean(PREF_KEY, false);
        } catch (Throwable t) {
            logError("Failed to read nearby share state", t);
            return false;
        }
    }

    private void setNearbyShareEnabled(Object tile, Context context, ClassLoader classLoader,
                                       String managerClass, String factoryMethod, String setMethod) {
        try {
            Class<?> mc = classLoader.loadClass(managerClass);
            Method lMethod = mc.getDeclaredMethod(factoryMethod, Context.class);
            Object manager = lMethod.invoke(null, context);
            if (manager == null) {
                log("Unable to get manager!");
                return;
            }
            Method zMethod = manager.getClass().getDeclaredMethod(setMethod, boolean.class);
            zMethod.setAccessible(true);
            zMethod.invoke(manager, true);

            // 同样动态查找 b() 方法
            String bMethodName = "b";
            DexKitBridge bridge = DexKitHelper.INSTANCE.getBridgeForClass(classLoader, TARGET_CLASS);
            if (bridge != null) {
                try {
                    MethodData md = bridge.findMethod(FindMethod.create()
                            .searchPackages(TARGET_PACKAGE)
                            .matcher(MethodMatcher.create()
                                    .paramTypes()
                                    .returnType("void")
                                    .declaredClass(TARGET_CLASS)
                            )
                    ).singleOrNull();
                    if (md != null) bMethodName = md.getName();
                } catch (Throwable ignored) {}
            }
            Method bMethod = findMethod(tile.getClass(), bMethodName);
            bMethod.setAccessible(true);
            bMethod.invoke(tile);

            log("successfully set share to enabled");
        } catch (Exception e) {
            logError("Failed to set nearby share to enable: ", e);
        }
    }
}

File: app/src/main/java/com/qimian233/ztool/hook/modules/mobiledesktop/BypassShareWarningHook.kt
package com.qimian233.ztool.hook.modules.mobiledesktop

import android.content.Context
import android.content.SharedPreferences
import com.qimian233.ztool.data.keys.ScopeKeys
import com.qimian233.ztool.dexindex.base.DexIndexConstants
import com.qimian233.ztool.hook.base.AppHookModule
import com.qimian233.ztool.hook.base.DexIndexStore
import io.github.libxposed.api.XposedModuleInterface
import androidx.core.content.edit

/**
 * Hook to bypass the share warning dialog of Super Interconnect (FileUnion).
 *
 * Uses DexKit to dynamically match the obfuscated method names by method
 * signature (parameter types + return type) instead of relying on hardcoded
 * single-letter names.
 */
class BypassShareWarningHook : AppHookModule() {

    companion object {
        private val TARGET_PACKAGE = arrayOf(ScopeKeys.MOBILE_DESKTOP.packageName)
        private const val TARGET_CLASS = "com.motorola.readyfor.tile.BaseFileUnionTile"
        private const val DIALOG_CLASS =
            "com.motorola.readyfor.common.dialog.ActionNoticeCommonDialogActivity"
        private const val MANAGER_PKG = "com.motorola.mobiledesktop.manager"
        private const val PREFS_NAME = "moto_ble_preference"
        private const val PREF_KEY1 = "file_union_transfer_switch"
        private const val PREF_KEY2 = "nearby_send_files"
    }

    override fun getModuleName(): String = "bypass_share_warning"

    override fun getTargetPackages(): Array<String> = TARGET_PACKAGE

    override fun handleLoadPackage(param: XposedModuleInterface.PackageLoadedParam) {
        val classLoader = param.defaultClassLoader

        // ── Read obfuscated class/method names from the offline index ────────
        val module = DexIndexStore.lookup(xposed, ScopeKeys.MOBILE_DESKTOP.packageName)
            ?.getAsJsonObject(DexIndexConstants.ModuleKeys.BYPASS_SHARE_WARNING)

        val managerClassName = module?.get(DexIndexConstants.Keys.MANAGER_CLASS)
            ?.takeIf { !it.isJsonNull }?.asString ?: "$MANAGER_PKG.c0"
        val managerFactoryMethodName = module?.get(DexIndexConstants.Keys.MANAGER_FACTORY_METHOD)
            ?.takeIf { !it.isJsonNull }?.asString ?: "l"   // static factory: (Context) → manager
        val managerSetMethodName = module?.get(DexIndexConstants.Keys.MANAGER_SET_METHOD)
            ?.takeIf { !it.isJsonNull }?.asString ?: "z"    // instance: (boolean) → void
        val tileRefreshMethod = module?.get(DexIndexConstants.Keys.TILE_REFRESH_METHOD)
            ?.takeIf { !it.isJsonNull }?.asString ?: "b"

        try {
            // ── Hook 1: tile click ───────────────────────────────────
            val baseFileUnionTileClass = classLoader.loadClass(TARGET_CLASS)
            val onClickMethod = baseFileUnionTileClass.getDeclaredMethod("onClick")
            hookWithId(onClickMethod, "on_click") { chain ->
                val tile = chain.thisObject
                val context = getContext(tile)
                if (context == null) {
                    chain.proceed()
                } else {
                    val enabled = isNearbyShareEnabled(context)
                    logger.debug("IsNearbyShareEnabled: $enabled")
                    if (enabled) {
                        logger.debug("Nearby share already enabled, keep original disable flow.")
                        chain.proceed()
                    } else {
                        setNearbyShareEnabled(
                            tile, context, classLoader,
                            managerClassName, managerFactoryMethodName, managerSetMethodName,
                            tileRefreshMethod
                        )
                        logger.debug("Bypassed warning and enabled nearby share directly.")
                        null
                    }
                }
            }
            logger.info("Installed hook for BaseFileUnionTile.onClick")
        } catch (t: Throwable) {
            logger.error("Failed to hook BaseFileUnionTile.onClick", t)
        }

        try {
            // ── Hook 2: generic dialog scenario ─────────────────────────
            val actionNoticeClass = classLoader.loadClass(DIALOG_CLASS)

            // The p() method name comes from the offline index — no-arg void + references the file_share_expose_title field
            val finalPMethodName = module?.get(DexIndexConstants.Keys.DIALOG_METHOD)
                ?.takeIf { !it.isJsonNull }?.asString ?: "t" // default fallback (current version)
            logger.debug("target method name of \"createAndStartExposureWarnDialog\": $finalPMethodName")

            val pMethod = actionNoticeClass.getDeclaredMethod(finalPMethodName)
            hookWithId(pMethod, "hook_162") {  chain ->
                val myObject = chain.thisObject
                val context = getContext(myObject)

                // Try enabling via the newer MotoDiscoveryManager
                try {
                    val qClass = classLoader.loadClass("com.motorola.motoaccount.sdk.gf.q")
                    val lMethod = qClass.getDeclaredMethod("l", Context::class.java)
                    val qInstance = lMethod.invoke(null, context)
                    qInstance.javaClass
                        .getDeclaredMethod("B", Boolean::class.javaPrimitiveType)
                        .invoke(qInstance, true)
                    logger.debug("dialog hook: enabled via MotoDiscoveryManager")
                } catch (_: ReflectiveOperationException) {
                    // Fall back to the legacy manager
                    val managerClass = classLoader.loadClass(managerClassName)
                    val lMethod =
                        managerClass.getDeclaredMethod(managerFactoryMethodName, Context::class.java)
                    val manager = lMethod.invoke(null, context)
                    if (manager == null) {
                        logger.warn("Unable to get manager!")
                    } else {
                        val zMethod = manager.javaClass.getDeclaredMethod(
                            managerSetMethodName, Boolean::class.javaPrimitiveType
                        )
                        zMethod.isAccessible = true
                        zMethod.invoke(manager, true)
                    }
                }
                null
            }
            logger.info("Installed hook for dialog activity method: $finalPMethodName")
        } catch (e: Exception) {
            logger.error("Failed to hook createAndStartExposureWarnIntent: ", e)
        }
    }

    private fun getContext(tile: Any?): Context? {
        return try {
            val getApplicationContextMethod = tile!!.javaClass.getMethod("getApplicationContext")
            val context = getApplicationContextMethod.invoke(tile)
            context as? Context
        } catch (_: Throwable) {
            null
        }
    }

    private fun isNearbyShareEnabled(context: Context): Boolean {
        val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        return try {
            prefs.getBoolean(PREF_KEY1, false) || prefs.getBoolean(PREF_KEY2, false)
        } catch (t: Throwable) {
            logger.error("Failed to read nearby share state", t)
            false
        }
    }

    private fun setNearbyShareEnabled(
        tile: Any?,
        context: Context,
        classLoader: ClassLoader,
        managerClass: String,
        factoryMethod: String,
        setMethod: String,
        tileRefreshMethod: String
    ) {
        // Try multiple enable strategies to support both old and new versions:
        //   Legacy: managerClass.l(context).z(true)  (c0)
        //   New:    q.l(context).B(true)             (MotoDiscoveryManager)
        //   Fallback: write SharedPreferences directly
        try {
            // ── Strategy 1: legacy manager class ──────────────────────────
            try {
                val mc = classLoader.loadClass(managerClass)
                val lMethod = mc.getDeclaredMethod(factoryMethod, Context::class.java)
                val manager = lMethod.invoke(null, context)
                if (manager != null) {
                    val zMethod = manager.javaClass.getDeclaredMethod(
                        setMethod, Boolean::class.javaPrimitiveType
                    )
                    zMethod.isAccessible = true
                    zMethod.invoke(manager, true)
                    logger.debug("enabled via legacy manager: $managerClass.$factoryMethod/$setMethod")
                }
            } catch (_: ReflectiveOperationException) {
                // ── Strategy 2: newer MotoDiscoveryManager ───────────────
                logger.warn("legacy manager not found, trying MotoDiscoveryManager")
                val qClass = classLoader.loadClass("com.motorola.motoaccount.sdk.gf.q")
                val lMethod = qClass.getDeclaredMethod("l", Context::class.java)
                val qInstance = lMethod.invoke(null, context)
                qInstance.javaClass
                    .getDeclaredMethod("B", Boolean::class.javaPrimitiveType)
                    .invoke(qInstance, true)
                logger.debug("enabled via MotoDiscoveryManager.q.l().B(true)")
            }

            // ── Fallback: write both SharedPreferences directly ───────────────
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
                .edit { putBoolean(PREF_KEY1, true) }
            context.getSharedPreferences("sp_file_ble", Context.MODE_PRIVATE)
                .edit { putBoolean("nearby_send_files", true) }

            // ── Refresh the tile via its b() method (resolved in handleLoadPackage) ────
            val bMethod = findMethod(tile!!.javaClass, tileRefreshMethod)
            bMethod.isAccessible = true
            bMethod.invoke(tile)

            logger.debug("successfully set share to enabled")
        } catch (e: Throwable) {
            logger.error("Failed to set nearby share to enable: ", e)
        }
    }
}
